src/vpn-api/worker.ts

```diff
@@ -71,91 +71,245 @@ const getInstanceSubnet = (instanceId: number) => {
+		let msg = `session ended for`;
+		if (data.username == null) {
+			return logger.debug(`${msg} rejected client_id=${clientId}`);
+		}
+		msg = `${msg} client_id=${clientId} uuid=${data.username}`;
+		if (
+			data.bytes_received != null &&
+			data.bytes_sent != null &&
+			data.time_duration > 0
+		) {
+			msg = `${msg} (len=${data.time_duration} rx=${data.bytes_received} tx=${
+				data.bytes_sent
+			})`;
+			metrics.histogram(Metrics.SessionRxBytes, data.bytes_received);
+			metrics.histogram(Metrics.SessionTxBytes, data.bytes_sent);
+			writeBitrateMetrics(data as VpnClientBytecountData, data.time_duration);
+		}
+		logger.info(msg);
+		metrics.dec(Metrics.OnlineDevices);
+		delete clientLog[clientId];
+		return clients.disconnected(data).then(logStateUpdate);
+	});
+
+	// setup metrics for prometheus
+	const kb = 2 ** 10; // 1024
+	const mb = 2 ** 10 * kb;
+	const gb = 2 ** 10 * mb;
+	const tb = 2 ** 10 * gb;
+	const bandwidthBuckets = [
+		mb,
+		10 * mb,
+		100 * mb,
+		500 * mb,
+		gb,
+		10 * gb,
+		100 * gb,
+		150 * gb,
+		250 * gb,
+		500 * gb,
+		tb,
+	];
+	const bitrateBuckets = [
+		100 * kb,
+		250 * kb,
+		500 * kb,
+		mb,
+		2 * mb,
+		3 * mb,
+		4 * mb,
+		5 * mb,
+		6 * mb,
+		7 * mb,
+		8 * mb,
+		9 * mb,
+		10 * mb,
+	];
+	metrics.describe(
+		Metrics.SessionRxBytes,
+		'histogram of rx bytes per vpn session',
+		{ buckets: bandwidthBuckets },
+	);
+	metrics.describe(
+		Metrics.SessionTxBytes,
+		'histogram of tx bytes per vpn session',
+		{ buckets: bandwidthBuckets },
+	);
+	metrics.describe(
+		Metrics.SessionRxBitrate,
+		'histogram of average rx rate per vpn session',
+		{ buckets: bitrateBuckets },
+	);
+	metrics.describe(
+		Metrics.SessionTxBitrate,
+		'histogram of average tx rate per vpn session',
+		{ buckets: bitrateBuckets },
+	);
+	metrics.describe(Metrics.OnlineDevices, 'vpn current online devices');
+	metrics.gauge(Metrics.OnlineDevices, 0);
+	metrics.describe(Metrics.TotalDevices, 'vpn total devices since restart');
+	metrics.counter(Metrics.TotalDevices, 0);
+	metrics.describe(
+		Metrics.AuthFailures,
+		'vpn device auth failures since restart',
+	);
+	metrics.counter(Metrics.AuthFailures, 0);
+
+	logger.notice(`starting...`);
+	return (
+		vpn
+			.start()
+			.tap(() => logger.info(`openvpn process started`))
+			// connect to vpn management console, setup logging & bytecount reporting, then release management hold
+			.bind(vpn)
+			.tap(vpn.connect)
+			.tap(vpn.enableLogging)
+			.tap(() => vpn.enableBytecountReporting(VPN_BYTECOUNT_INTERVAL))
+			.tap(vpn.releaseHold)
+			.tap(() => logger.info(`management hold released`))
+			// register as haproxy backend
+			.tap(() =>
+				new HAProxy('/var/run/haproxy.sock').register(
+					`vpn-workers/vpn${instanceId}`,
+					vpnPort,
+				),
+			)
+			.tap(() =>
+				logger.info(
+					`registered as haproxy backend server vpn-workers/vpn${instanceId}`,
+				),
+			)
+			.catch(fatalErrorHandler)
+			.tap(() => logger.notice(`waiting for clients...`))
```

Comment:
You've flagged this pattern a few time (`.tap(() => blah())`) showing a preference for `.tap(() => {blah()})`.

Is this just a preference or is there a technical benefit?

Reply:
It makes it clear to the reader that no value is ever returned which is the reason for the preference, but it also does the same for v8 which may allow it to make optimisations that it otherwise couldn't (whether or not it actually uses that info is another matter, and whether in some of these cases it can infer it regardless of the less clear pattern I cannot say though - but providing the info gives it the best chance).  Also in some cases returning a value that we don't care about may still end up in it getting used/stored, eg returning an object to bluebird will require it to do additional checks to see if it's a promise that it should be waiting on and if it was a `.then` instead of a `.tap` it has to store that value and pass it on as it doesn't know that it's actually unused.  Generally it's good practice imo to not return anything if you don't want it to be acted upon rather than relying on the current implementation details meaning that it won't be as those could change.

That said regardless of the technical reasons I'd still want this pattern because of the clarity to people reading the code rather than potentially having to double check that the return value actually isn't important

Reply:
Thanks!